consumer/pdatautil/metricsbuilder.go

```diff
@@ -0,0 +1,205 @@
+// metric/label combination.
+//
+// It is NOT thread-safe, so you should use an external mutex if using it from
+// multiple goroutines.
+type MetricsBuilder struct {
+	// If set, this instrumentation library name will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryName string
+	// If set, this instrumentation library version will be used for all metrics
+	// generated by this builder.
+	InstrumentationLibraryVersion string
+	// These attributes will be added to the Resource object on all
+	// ResourceMetrics instances created by the builder.
+	ResourceAttributes map[string]string
+	// This time will be used as the Timestamp for all metrics generated.  It
+	// can be updated with a new timestamp at any time.
+	Timestamp time.Time
+
+	metrics pdata.MetricSlice
+	labels  map[string]string
+	isInit  bool
+}
+
+// Flush creates a pdata.Metrics instance with all of the pending metrics on
+// the builder and returns it.  Subsequent calls to Flush will only return
+// metrics added after the previous Flush call.  Labels set with With*Labels
+// will be retained after a call to Flush.
+func (mb *MetricsBuilder) Flush() pdata.Metrics {
+	md := data.NewMetricData()
+	rms := md.ResourceMetrics()
+	rms.Resize(1)
+	rm := rms.At(0)
+	rm.InitEmpty()
+
+	res := rm.Resource()
+	res.InitEmpty()
+	for k, v := range mb.ResourceAttributes {
+		res.Attributes().Insert(k, pdata.NewAttributeValueString(v))
+	}
+
+	ilms := rm.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	ilm := ilms.At(0)
+	ilm.InitEmpty()
+
+	il := ilm.InstrumentationLibrary()
+	il.InitEmpty()
+	il.SetName(mb.InstrumentationLibraryName)
+	il.SetVersion(mb.InstrumentationLibraryVersion)
+
+	mb.metrics.MoveAndAppendTo(ilm.Metrics())
+
+	mb.init()
+
+	return pdata.Metrics{InternalOpaque: md}
+}
+
+func (mb *MetricsBuilder) init() {
+	mb.metrics = pdata.NewMetricSlice()
+	mb.isInit = true
+}
+
+// WithLabels sets a set of labels on the builder. All subsequent calls to
+// create metrics will create metrics that use the most recently set of labels
+// set with this method.  The input map's entries are copied so the map can be
+// mutated freely without affecting the builder.
+func (mb *MetricsBuilder) WithLabels(l map[string]string) *MetricsBuilder {
+	mb.labels = nil
+	return mb.WithAdditionalLabels(l)
+}
+
+// WithAdditionalLabels will add the given labels to the already existing label
+// set, instead of replacing them outright.  The input map's entries are copied
+// so the map can be mutated freely without affecting the builder.
+func (mb *MetricsBuilder) WithAdditionalLabels(l map[string]string) *MetricsBuilder {
+	if mb.labels == nil {
+		mb.labels = map[string]string{}
+	}
+	for k, v := range l {
+		mb.labels[k] = v
+	}
+	return mb
+}
+
+func (mb *MetricsBuilder) appendMetric(m *pdata.Metric) {
+	if !mb.isInit {
+		mb.init()
+	}
+	mb.metrics.Append(m)
+}
+
+func (mb *MetricsBuilder) AddGauge(name string, metricValue int64) *MetricsBuilder {
```

Comment:
I think adopting otel metric terminology would be preferable for the project:
`AddGauge` -> `AddInt64UpDownCounter`,
`AddGaugeD` ->`AddFloat64UpDownCounter`,
`AddCumulative` -> `AddInt64Counter`,
`AddCumulativeD` -> `AddFloat64Counter`

Reply:
I have never seen the term `UpDownCounter` before but it just seems like a very peculiar way to say "gauge", which has a long established usage across many monitoring systems.

Also neither of these match the metric type names in `pdata` (see `metrics.go`).  For cumulative they are called `monotonic` in pdata, but that is more of a mathematical property of them whereas "cumulative" better conveys the semantics in an observability context.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#metric-instruments

Reply:
If you wouldn't want to adopt the terminology of the specification for instruments, I'd at least use the metric types since gauge and cumulative don't appear to be adopted by otel.

Reply:
OTLP was updated and uses "gauge" terminology: https://github.com/open-telemetry/opentelemetry-proto/blob/313a868be259dce6c6516dd417d3ad5fd3321acf/opentelemetry/proto/metrics/v1/metrics.proto#L139

Reply:
The builder now uses the new standard terms in the final Metrics model.